File: hacp-ts/src/crypto.ts
import * as crypto from "crypto";

// DER SPKI prefix for a raw 32-byte Ed25519 public key.
const SPKI_PREFIX = Buffer.from("302a300506032b6570032100", "hex");

export function sha256Hex(data: Buffer): string {
  return crypto.createHash("sha256").update(data).digest("hex");
}

export function loadPublicKey(hexKey: string): crypto.KeyObject {
  const raw = Buffer.from(hexKey, "hex");
  if (raw.length !== 32) {
    throw new Error(`public key must be 32 bytes, got ${raw.length}`);
  }
  const spki = Buffer.concat([SPKI_PREFIX, raw]);
  return crypto.createPublicKey({ key: spki, format: "der", type: "spki" });
}

export function verifySignature(
  pub: crypto.KeyObject,
  payload: Buffer,
  sigB64Url: string
): boolean {
  try {
    const sig = Buffer.from(sigB64Url, "base64url");
    return crypto.verify(null, payload, pub, sig);
  } catch {
    return false;
  }
}

File: hacp-go/crypto.go
package main

import (
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"fmt"
)

// SHA256Hex returns lowercase hex SHA-256 digest.
func SHA256Hex(data []byte) string {
	h := sha256.Sum256(data)
	return hex.EncodeToString(h[:])
}

// LoadPublicKey parses a 32-byte raw Ed25519 public key from hex.
func LoadPublicKey(hexKey string) (ed25519.PublicKey, error) {
	raw, err := hex.DecodeString(hexKey)
	if err != nil {
		return nil, err
	}
	if len(raw) != ed25519.PublicKeySize {
		return nil, fmt.Errorf("public key must be %d bytes, got %d", ed25519.PublicKeySize, len(raw))
	}
	return ed25519.PublicKey(raw), nil
}

// VerifySignature checks a base64url (no padding) Ed25519 signature.
func VerifySignature(pub ed25519.PublicKey, payload []byte, sigB64 string) bool {
	sig, err := base64.RawURLEncoding.DecodeString(sigB64)
	if err != nil {
		return false
	}
	return ed25519.Verify(pub, payload, sig)
}